kube/client.go

```diff
@@ -26,26 +26,27 @@ import (
 	"tailscale.com/util/multierr"
 )
 
+// defaultURL is the default URL to use for the Kubernetes API.
+// saPath is the path to the service account directory.
+// Their values are set by the init function.
+// rootPathForTests is set by tests to override the root path to the
+// service account directory.
 var (
 	saPath     string
 	defaultURL string
+	rootPathForTests string
 )
 
-// rootPathForTests is set by tests to override the root path to the
-// service account directory.
-var rootPathForTests string
-
-// Allow overriding kube SA path and default URL via environment variables.
+// init sets the default URL and service account path.
 func init() {
-	saPath = getEnvOrDefault("KUBE_SA_PATH", "/var/run/secrets/kubernetes.io/serviceaccount")
-	defaultURL = getEnvOrDefault("KUBE_DEFAULT_URL", "https://kubernetes.default.svc")
-}
-
-func getEnvOrDefault(key, defaultValue string) string {
-	if value, exists := os.LookupEnv(key); exists {
-		return value
+	if os.Getenv("TS_KUBERNETES_READ_API_SERVER_ADDRESS_FROM_ENV") != "" {
+		host, port := os.Getenv("KUBERNETES_SERVICE_HOST"), os.Getenv("KUBERNETES_SERVICE_PORT")
+		defaultURL = "https://" + net.JoinHostPort(host, port)
+	} else {
+		defaultURL = "https://kubernetes.default.svc"
 	}
-	return defaultValue
+
+	saPath = "/var/run/secrets/kubernetes.io/serviceaccount"
```

Comment:
> I do feel like some of this would be simpler using client-go

It definitely would, the reason why we don't do that is to keep the binary size smaller (the container image is used not just on kube)

> I feel that the simplest solution is to use SetURL for this as well: [ab5fbed](https://github.com/tailscale/tailscale/pull/12115/commits/ab5fbed6771373bd7b25d467b7f9f0aaf192622e)

There is another kube client  that gets initiated when state storage is a Kubernetes `Secret`, [here](https://github.com/tailscale/tailscale/blob/v1.67.0-pre/ipn/store/kubestore/store_kube.go#L29) to read/update the `Secret`. Your PR now only updates the client that is initiated to check the `Secret` permissions, whilst the client for state store will still use the default URL. (This code _is_ a little bit convoluted as it has grown naturally and at some point will need refactor, but right now we are low bandwidth).

Reply:
Thanks for the updates @ChandonPierre .

> still not very DRY I'm afraid

I think this is the best we can get without a bigger refactor of the state.

Happy for this one to go in just one more nit- can we change `os.Getenv("TS_KUBERNETES_READ_API_SERVER_ADDRESS_FROM_ENV") != ""` -> `os.Getenv("TS_KUBERNETES_READ_API_SERVER_ADDRESS_FROM_ENV") == "true"` or something like that to not have folks wonder what value needs to be used & how to unset?